Sources/Datadog/DDTracer.swift

```diff
@@ -0,0 +1,60 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-2020 Datadog, Inc.
+ */
+
+import OpenTracing
+import Foundation
+
+public class DDTracer: Tracer {
+    /// Writes `Span` objects to output.
+    private let spanOutput: SpanOutput
+
+    // TODO: RUMM-332 Consider builder pattern to initialize the tracer
+    internal init(spanOutput: SpanOutput) {
+        self.spanOutput = spanOutput
+    }
+
+    // MARK: - Open Tracing interface
+
+    public func startSpan(operationName: String, references: [Reference]? = nil, tags: [String: Codable]? = nil, startTime: Date? = nil) -> Span {
+        do {
+            return try startSpanOrThrow(operationName: operationName, references: references, tags: tags, startTime: startTime)
+        } catch {
+            consolePrint("🔥 \(error)")
+            return DDNoopSpan()
+        }
+    }
+
+    public func inject(spanContext: SpanContext, writer: FormatWriter) {
+        // TODO: RUMM-292
+    }
+
+    public func extract(reader: FormatReader) -> SpanContext? {
+        // TODO: RUMM-292
+        return nil
+    }
+
+    // MARK: - Private Open Tracing helpers
+
+    private func startSpanOrThrow(operationName: String, references: [Reference]?, tags: [String: Codable]?, startTime: Date?) throws -> Span {
+        guard let datadog = Datadog.instance else {
+            throw ProgrammerError(description: "`Datadog.initialize()` must be called prior to `startSpan(...)`.")
+        }
+
+        let parentSpanContext = references?.compactMap { $0.context as? DDSpanContext }.last
+        return DDSpan(
+            tracer: self,
+            operationName: operationName,
+            parentSpanContext: parentSpanContext,
+            startTime: startTime ?? datadog.dateProvider.currentDate()
```

Comment:
If the default behavior is basically `Date()`, how about setting the default parameter to `Date()`?
Please correct me if I'm wrong: currently the default parameter is `nil` but actual behavior is closer to `Date()`

And we still can pass `datadog.dateProvider.currentDate()` in unit tests

Reply:
`datadog.dateProvider.currentDate()` acts as a `Date()` mock for us. By hiding it inside `dateProvider`, we are able to mock the creation date (here for `Span`, but in logs we do it for `Log`), so it's testable and **test-environment agnostic** (runs the same no matter of time zone or current wall time):

```swift
let dateProvider = RelativeDateProvider(using: .mockDecember15th2019At10AMUTC())
Datadog.instance = .mockNoOpWith(dateProvider: dateProvider)

let span = tracer.startSpan(operationName: "operation")

XCTAssertEqual(recorded?.span.startTime, .mockDecember15th2019At10AMUTC())
```

If we default to `Date()` here, the unit tests all around would have have to be defined based on "time difference" assumption:
```swift
let span = tracer.startSpan(operationName: "operation")

// Here we'd asume that `0.5s` is enough time for the tests runner to execute first line.
XCTAssertTrue(recorded?.span.startTime.isNoOlderThan(seconds: 0.5))
```
That's a bit odd and will not work if we want to mock, let's say, 3 spans which start every `1s`. The only way would be:
```swift
let span1 = tracer.startSpan(operationName: "operation 1")
Thread.sleep(1)
let span2 = tracer.startSpan(operationName: "operation 1")
Thread.sleep(1)
let span3 = tracer.startSpan(operationName: "operation 1")
```

The `RelativeDateProvider` is much more flexible and supports this:
```swift
RelativeDateProvider(
   startingFrom: .mockDecember15th2019At10AMUTC(),
   advancingBySeconds: 1
)

let span1 = tracer.startSpan(operationName: "op 1")
let span2 = tracer.startSpan(operationName: "op 1") // startDate = span1.startDate + 1
let span3 = tracer.startSpan(operationName: "op 1") // startDate = span1.startDate + 2
```

This approach was extremely handy for logs, where some of them need to be written "X seconds ago" so that data upload worker considers them valid for upload. Same will happen in tracing at some point.

Reply:
Sorry, my comment was not clear enough.

Sure, `dateProvider` is essential for unit tests. 
Here is what I'm proposing:
```swift
// DDTracer.swift
func startSpan(..., startTime: Date = Date()) { ... }

// TracerTests.swift
let mockDateProvider = MockDateProvider()
tracer.startSpan(..., startTime = mockDateProvider.time)
```
The default behavior in real-world use cases will be explicit in method signature while we keep mocking the value in our unit tests.
• real-world default behavior -> default behavior 
• unit test behavior -> non-default behavior

Reply:
Ah, I see what you mean, thanks for clarifying 👌.

The difference is that when putting `startTime: Date = Date()` as default, we need a direct access to this method in unit tests to provide the mock (`mockDateProvider.currenDate()`). This works 👍 as long as we have it (like in `DDTracerTests`), but won't work if we need to wrap `startSpan(...)` into another component which we want to unit test.

One example which I can think of now is `URLSessionInterceptor` (like we have [`TracingInterceptor`](https://github.com/DataDog/dd-sdk-android/blob/386a5db15c321e3dc93afab8e236d5420723cab1/dd-sdk-android/src/main/kotlin/com/datadog/android/tracing/TracingInterceptor.kt#L53) in Android SDK). It will use `tracer.startSpan(...)` internally, so we won't have ability to provide the date mock.

This is exactly why global `datadog.dateProvider` come super handy in logging part, where log date creation happens deeply inside `LogBuilder`, which is used by `LogOutput` which is used by `Logger`. By mocking `datadog.dateProvider` we are able to mock the value for `log.date` written to disk in `LoggerTests`.

And last thing - in production code, `dateProvider.currentDate()` is in fact evaluated to `Date()` as this function is inlined by the compiler:
```swift
// DateProvider.swift

internal struct SystemDateProvider: DateProvider {
    @inlinable
    func currentDate() -> Date { return Date() }
}
```

Reply:
Okay I see 👌 
My concern was clarity of the method signature (ie: avoiding developer questions such as "I passed `nil`, why does that date come out?")
But I see its pros outweigh its cons 👍 

that `@inlinable` is 💯 👏

Reply:
Ah, speaking of `startTime: Date` vs `startTime: Date?` we don't control that - it's part of `OpenTracing.Tracer` interface and specification that the `startTime` parameter is optional and if avoided, wall time should be used. And I'm happy they choosed `Date? = nil` over `Date = Date()` as this lets us to instal this internal mock.